examples/shell/cmd_otcli.cpp

```diff
@@ -0,0 +1,168 @@
+
+#include <stdio.h>
+
+#include <platform/ThreadStackManager.h>
+#include <shell/shell.h>
+#include <support/CHIPArgParser.hpp>
+#include <support/CHIPMem.h>
+#include <support/CodeUtils.h>
+
+#if CHIP_TARGET_STYLE_EMBEDDED
+#include <openthread/cli.h>
+#include <openthread/instance.h>
+#include <openthread/ip6.h>
+#include <openthread/link.h>
+#include <openthread/thread.h>
+#else
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
+#endif
+
+using namespace chip;
+using namespace chip::Shell;
+using namespace chip::Platform;
+using namespace chip::DeviceLayer;
+using namespace chip::Logging;
+using namespace chip::ArgParser;
+
+static chip::Shell::Shell sShellOtcliSubcommands;
+
+int cmd_otcli_help_iterator(shell_command_t * command, void * arg)
+{
+    streamer_printf(streamer_get(), "  %-15s %s\n\r", command->cmd_name, command->cmd_help);
+    return 0;
+}
+
+int cmd_otcli_help(int argc, char ** argv)
+{
+    sShellOtcliSubcommands.ForEachCommand(cmd_otcli_help_iterator, NULL);
+    return 0;
+}
+
+#if CHIP_TARGET_STYLE_EMBEDDED
+
+int cmd_otcli_dispatch(int argc, char ** argv)
+{
+    CHIP_ERROR error = CHIP_NO_ERROR;
+
+// From OT CLI internal lib, kMaxLineLength = 128
+#define kMaxLineLength 128
+    char buff[kMaxLineLength] = { 0 };
+    char * buff_ptr           = buff;
+    int i                     = 0;
+
+    VerifyOrExit(argc > 0, error = CHIP_ERROR_INVALID_ARGUMENT);
+
+    for (i = 0; i < argc; i++)
+    {
+        size_t arg_len = strlen(argv[i]);
+
+        /* Make sure that the next argument won't overflow the buffer */
+        VerifyOrExit(buff_ptr + arg_len < buff + kMaxLineLength, error = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+        strncpy(buff_ptr, argv[i], arg_len);
+        buff_ptr += arg_len;
+
+        /* Make sure that there is enough buffer for a space char */
+        if (buff_ptr + sizeof(char) < buff + kMaxLineLength)
+        {
+            strncpy(buff_ptr, " ", sizeof(char));
+            buff_ptr++;
+        }
+    }
+    buff_ptr = 0;
+
+    otCliConsoleInputLine(buff, buff_ptr - buff);
+exit:
+    return error;
+}
+
+#elif CHIP_TARGET_STYLE_UNIX
+
+int cmd_otcli_dispatch(int argc, char ** argv)
+{
+    CHIP_ERROR error = CHIP_NO_ERROR;
+
+    int pid, status;
+    uid_t euid         = geteuid();
+    char ctl_command[] = "/usr/local/sbin/ot-ctl";
+
+    // Must run as sudo.
+    if (euid != 0)
+    {
+        streamer_printf(streamer_get(), "Error otcli: requires running chip-shell as sudo\n\r");
+        error = CHIP_ERROR_INCORRECT_STATE;
+        ExitNow();
+    }
+
+    VerifyOrExit(argc > 0, error = CHIP_ERROR_INVALID_ARGUMENT);
+
+    // Fork and execute the command.
+    pid = fork();
+    VerifyOrExit(pid != -1, error = CHIP_ERROR_INCORRECT_STATE);
+
+    if (pid == 0)
+    {
+        // Child process to execute the command with provided arguments
+        --argv; // Restore access to entry [0] containing the command;
+        argv[0] = ctl_command;
+        execv(ctl_command, argv);
```

Comment:
Please check the return value, e.g.

```
if (execve() < 0) {
  fprintf(stderr, "execv: %s: %s\n", ctl_command[0], strerror(errno));
  exit(1);
}
```

Reply:
Improved error handling and passing from child to parent.